framework/src/userobject/ElementQualityChecker.C

```diff
@@ -0,0 +1,201 @@
+    _lower_bound(_has_lower_bound ? getParam<Real>("lower_bound") : 0.0),
+    _failure_type(getParam<MooseEnum>("failure_type").getEnum<FailureType>())
+{
+  if (_has_lower_bound && _has_upper_bound && _lower_bound >= _upper_bound)
+    mooseError("Provided lower bound should be less than provided upper bound!");
+}
+
+void
+ElementQualityChecker::initialize()
+{
+  _m_values.clear();
+  _elem_ids.clear();
+  _bypassed = 0;
+}
+
+void
+ElementQualityChecker::execute()
+{
+  // obtain the available quality metric for current ElemType
+  std::vector<ElemQuality> metrics_avail = libMesh::Quality::valid(_current_elem->type());
+
+  // check whether the provided quality metric is applicable to current ElemType
+  if (!checkMetricApplicability(_m_type, metrics_avail))
+  {
+    _bypassed = 1;
+    return;
+  }
+
+  std::pair<Real, Real> default_bounds = _current_elem->qual_bounds(_m_type);
+  std::pair<Real, Real> actual_bounds;
+  if (_has_lower_bound && _has_upper_bound)
+    actual_bounds = std::make_pair(_lower_bound, _upper_bound);
+  else if (_has_lower_bound)
+  {
+    if (_lower_bound < default_bounds.second)
+      actual_bounds = std::make_pair(_lower_bound, default_bounds.second);
+    else
+      mooseError("Provided lower bound should less than the default upper bound: ",
+                 default_bounds.second);
+  }
+  else if (_has_upper_bound)
+  {
+    if (_upper_bound > default_bounds.first)
+      actual_bounds = std::make_pair(default_bounds.first, _upper_bound);
+    else
+      mooseError("Provided upper bound should larger than the default lower bound: ",
+                 default_bounds.first);
+  }
+  else
+    actual_bounds = default_bounds;
+
+  // calculate and save quality metric value for current element
+  Real mv = _current_elem->quality(_m_type);
+  _m_values.insert(mv);
+
+  // check element quality metric, save ids of elements whose quality metrics exceeds the preset
+  // bounds
+  if (mv < actual_bounds.first || mv > actual_bounds.second)
+    _elem_ids.insert(_current_elem->id());
+}
+
+void
+ElementQualityChecker::threadJoin(const UserObject & uo)
+{
+  const ElementQualityChecker & eqc = static_cast<const ElementQualityChecker &>(uo);
+  _m_values.insert(eqc._m_values.begin(), eqc._m_values.end());
+  _elem_ids.insert(eqc._elem_ids.begin(), eqc._elem_ids.end());
+  _bypassed += eqc._bypassed;
+}
+
+void
+ElementQualityChecker::finalize()
+{
+  _communicator.set_union(_m_values);
+  _communicator.set_union(_elem_ids);
+  _communicator.sum(_bypassed);
+
+  if (_bypassed)
+  {
+    switch (_failure_type)
+    {
+      case FailureType::WARNING:
+        mooseWarning("Provided quality metric doesn't apply to certain element type!");
+        break;
+
+      case FailureType::ERROR:
+        mooseError("Provided quality metric doesn't apply to certain element type!");
+        break;
+
+      default:
+        mooseError("Unknown failure type!");
+    }
+  }
```

Comment:
So this is up to you if you really want to control element metrics separately from everything else in MOOSE, but we have another mechanism you can consider. `mooseWarning` already gets promoted to `mooseError` when you pass `--error` on the command line. We use that in our test suite and in many other places to trigger errors when warnings are printed. We created that system to avoid this very kind of logic here. It's all built-in so you can remove dozens of lines of code.

Reply:
This was my idea. We are planning to have an Action that sets up a few of these quality checks by default for an application code. The intent here was that we would have the Action set up two copies of some of these objects: one with a warning threshold and one with an error threshold (to completely refuse to run with more eggregious violations of the quality checks). Since we're trying to stop users from running with bad meshes, I don't think that relying them to use '--error' in this case makes sense.